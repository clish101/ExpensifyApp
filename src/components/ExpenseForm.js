import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/initialize';

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      description: props.expense ? props.expense.description : '',
      notes: props.expense ? props.expense.note : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: '',
    };
  }

  changeHandler = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };

  handleText = (e) => {
    const notes = e.target.value;
    this.setState(() => ({ notes }));
  };

  onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };

  onSubmit = (e) => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: 'Please provide description and amount' }));
    } else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.notes,
      });
    }
  };

  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            placeholder="Description"
            type="text"
            autoFocus
            value={this.state.description}
            onChange={this.changeHandler}
            name="describe"
          />
          <input
            type="number"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            placeholder="Notes for your expense(optional)"
            style={{ marginBottom: '-7px' }}
            value={this.state.notes}
            onChange={this.handleText}
          ></textarea>
          <button>Add Expense</button>
        </form>
      </div>
    );
  }
}
