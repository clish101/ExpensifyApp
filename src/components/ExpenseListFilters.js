import React from 'react';
import { connect } from 'react-redux';
import {
  setTextFilter,
  setStartDate,
  setEndDate,
  sortByDate,
  sortByAmount,
} from '../actions/filters';
import { DateRangePicker } from 'react-dates';

export class ExpenseListFilters extends React.Component {
  state = {
    calendarFocused: null,
  };

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onFocusChange = (focusedInput) => {
    this.setState(() => ({ calendarFocused: focusedInput }));
  };

  onTextChange = (e) => {
    this.props.setTextFilter(e.target.value);
  };

  onSortChange = (e) => {
    if (e.target.value === 'amount') {
      this.props.sortByAmount();
    } else {
      this.props.sortByDate();
    }
  };

  render() {
    return (
      <div>
        <select onChange={this.onSortChange}>
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <input value={this.props.filters.text} onChange={this.onTextChange} />
        <DateRangePicker
          startDate={this.props.filters.startDate}
          startDateId={'startDateId'}
          endDate={this.props.filters.endDate}
          endDateId={'endDateId'}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          isOutsideRange={() => false}
          showClearDates={true}
        />
      </div>
    );
  }
}

const MapDispatchToProps = (dispatch) => ({
  setTextFilter: (text) => dispatch(setTextFilter(text)),
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate)),
  sortByAmount: () => dispatch(sortByAmount()),
  sortByDate: () => dispatch(sortByDate()),
});

const MapStateToProps = (state) => {
  return {
    filters: state.filters,
  };
};

export default connect(MapStateToProps, MapDispatchToProps)(ExpenseListFilters);
