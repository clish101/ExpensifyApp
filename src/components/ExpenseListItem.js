import React from 'react';

import { Link } from 'react-router-dom';

const ExpenseListItem = ({ description, amount, createdAt, dispatch, id }) => (
  <div>
    <p>These are what was spend on</p>
    <Link to={`/edit/${id}`}>
      <h3>{description}</h3>
    </Link>
    amount - createdAt
    <p>
      {amount} - {createdAt}
    </p>
  </div>
);

export default ExpenseListItem;
