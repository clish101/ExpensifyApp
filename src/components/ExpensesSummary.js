import React from 'react';
import { connect } from 'react-redux';
import numeral from 'numeral';
import expenses from '../tests/fixtures/expenses';
import getFilteredExpenses from '../selectors/expenses_selector';
import expensesTotal from '../selectors/expenses-total';

export const ExpensesSummary = (props) => {
  return (
    <div>
      {props.expenseCount > 0 ? (
        <h1>
          Viewing {props.expenseCount}{' '}
          {props.expenseCount === 1 ? 'expense' : 'expenses'} totalling KES
          {numeral(props.expensesTotal).format('0,00.00')}
        </h1>
      ) : (
        <p>No Expenses to show</p>
      )}
    </div>
  );
};

const MapStateToProps = (state) => ({
  expenses: getFilteredExpenses(state.expenses, state.filters),
  expenseCount: getFilteredExpenses(state.expenses, state.filters).length,
  expensesTotal: getFilteredExpenses(state.expenses, state.filters)
    .map((x) => x.amount / 100)
    .reduce((a, b) => a + b, 0),
});

export default connect(MapStateToProps)(ExpensesSummary);
