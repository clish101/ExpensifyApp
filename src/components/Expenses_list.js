import React from 'react';
import { connect } from 'react-redux';
import getFilteredExpenses from '../selectors/expenses_selector';
import ExpenseListItem from './ExpenseListItem';

export const ExpenseList = (props) => (
  <div>
    {props.expenses.length === 0 ? (
      <p>No expenses to render</p>
    ) : (
      props.expenses.map((expense) => {
        return <ExpenseListItem key={expense.id} {...expense} />;
      })
    )}
  </div>
);

const MapStateToProps = (state) => {
  return {
    expenses: getFilteredExpenses(state.expenses, state.filters),
  };
};

export default connect(MapStateToProps)(ExpenseList);
