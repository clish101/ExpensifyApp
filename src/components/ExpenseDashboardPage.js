import React from 'react';
import ExpenseListFilters from './ExpenseListFilters';
import ExpensesSummary from './ExpensesSummary';
import Expenses_list from './Expenses_list';

const ExpenseDashboardPage = () => (
  <div>
    This is from my dashboard component
    <ExpensesSummary />
    <ExpenseListFilters />
    <Expenses_list />
  </div>
);

export default ExpenseDashboardPage;
