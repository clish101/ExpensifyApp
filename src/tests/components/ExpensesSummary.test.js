import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';
import React from 'react';
import getFilteredExpenses from '../../selectors/expenses_selector';
import { shallow } from 'enzyme';

test('Should set the data for expenses filtered by text', () => {
  const filters = {
    text: 'e',
    sortBy: 'date',
    startDate: undefined,
    endDate: undefined,
  };
  const wrapper = shallow(
    <ExpensesSummary
      expenseCount={getFilteredExpenses(expenses, filters).length}
      expensesTotal={getFilteredExpenses(expenses, filters)
        .map((x) => x.amount / 100)
        .reduce((a, b) => a + b, 0)}
    />
  );
  expect(wrapper).toMatchSnapshot();
});

test('Should set the data for "expense"  one expense that is', () => {
  const filters = {
    text: 'gu',
    sortBy: 'date',
    startDate: undefined,
    endDate: undefined,
  };
  const wrapper = shallow(
    <ExpensesSummary
      expenseCount={getFilteredExpenses(expenses, filters).length}
      expensesTotal={getFilteredExpenses(expenses, filters)
        .map((x) => x.amount / 100)
        .reduce((a, b) => a + b, 0)}
    />
  );
  expect(wrapper).toMatchSnapshot();
});
