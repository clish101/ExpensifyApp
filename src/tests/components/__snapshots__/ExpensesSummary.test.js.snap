// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should set the data for "expense"  one expense that is 1`] = `
<div>
  <p>
    Viewing 
    1
     
    expense
     totalling KES
    1.95
  </p>
</div>
`;

exports[`Should set the data for expenses filtered by text 1`] = `
<div>
  <p>
    Viewing 
    2
     
    expenses
     totalling KES
    1,140.00
  </p>
</div>
`;
